test/unittest/tests/test_futex.cpp

```diff
@@ -0,0 +1,313 @@
+    {
+        futex_atomic_store(&_futex, expected);
+        for (auto &w : _waiters)
+            _threads.push_back(thread(&waiter::wait_on, &w, &_futex, timeout_usec));
+    }
+
+    void
+    delay_then_wake(time_t wake_start_delay_usec, const vector<wake_signal> &wake_signals)
+    {
+        chrono::duration<time_t, micro> WAKE_DELAY{wake_start_delay_usec};
+        this_thread::sleep_for(WAKE_DELAY);
+        for (auto &&sig : wake_signals) {
+            REQUIRE(__wt_futex_wake(&_futex, sig._type, sig._value) == 0);
+            this_thread::sleep_for(INTER_WAKE_DELAY);
+        }
+    }
+
+    void
+    wait_and_check(const vector<wake_signal> &wake_signals)
+    {
+        size_t expected_timeouts;
+
+        for (auto &&t : _threads)
+            t.join();
+        CAPTURE(_waiters);
+
+        /* Infer the number of expected timeouts. */
+        if (wake_signals.empty())
+            expected_timeouts = _threads.size();
+        else if (wake_signals.begin()->_type == WT_FUTEX_WAKE_ALL)
+            expected_timeouts = 0;
+        else
+            expected_timeouts = _threads.size() - wake_signals.size();
+
+        auto result = check_outcomes(expected_timeouts, wake_signals);
+        CAPTURE(result);
+        REQUIRE((result == outcome::Expected || result == outcome::Acceptable));
+    }
+
+    /*
+     * check_outcomes --
+     *     Check completed waiters to see if they meet expectations.
+     *
+     * There must be either a WakeAll signal or a WakeOne signal for every waiter that is expected
+     *     to be awoken.
+     */
+    outcome
+    check_outcomes(const size_t timeouts, const vector<wake_signal> &wake_sigs)
+    {
+        list<wake_signal> wakes(wake_sigs.begin(), wake_sigs.end());
+        size_t timedout = 0;
+
+        /*
+         * Loop over the waiters and inspect their state.
+         *
+         * Note: spurious wakeups (no error and expected == wake up value) are detected after the
+         * loop exits so there are is no test in the loop.
+         */
+        for (auto &&w : _waiters) {
+            if (w._ret == -1) {
+                if (w._eno != ETIMEDOUT)
+                    return outcome::Error;
+                else if (w._val_on_wake != w._expected)
+                    return outcome::TimeoutWithUnexpectedValue;
+                timedout++;
+            } else if (w._eno != 0)
+                return outcome::RetZeroWithErrno;
+            else if (wakes.empty() && w._val_on_wake != w._expected)
+                return outcome::UnexpectWakeNonSpuriousWakeup;
+            else if (w._val_on_wake != w._expected) {
+                /* This is a non-spurious wake up. */
+                auto match = find_if(wakes.begin(), wakes.end(),
+                  [&w](const wake_signal &s) { return s._value == w._val_on_wake; });
+                if (match == wakes.end())
+                    return outcome::WakeWithUnexpectedValue;
+                if (match->_type == WT_FUTEX_WAKE_ONE)
```

Comment:
How is this check handling `WT_FUTEX_WAKE_ALL`? It seems to me that by the end of this loop, `wakes` would be non-empty, and we would return `outcome::Acceptable` and never `outcome::Expected`, but maybe I'm missing something.

Reply:
Reworked.